packages/test-sync/src/androidAndroidTest/kotlin/io/realm/kotlin/test/mongodb/shared/serializer/BsonEncoderTests.kt

```diff
@@ -0,0 +1,311 @@
+import kotlinx.serialization.serializer
+import org.mongodb.kbson.BsonArray
+import org.mongodb.kbson.BsonBinary
+import org.mongodb.kbson.BsonBinarySubType
+import org.mongodb.kbson.BsonBoolean
+import org.mongodb.kbson.BsonDateTime
+import org.mongodb.kbson.BsonDocument
+import org.mongodb.kbson.BsonDouble
+import org.mongodb.kbson.BsonInt32
+import org.mongodb.kbson.BsonInt64
+import org.mongodb.kbson.BsonInvalidOperationException
+import org.mongodb.kbson.BsonNull
+import org.mongodb.kbson.BsonNumber
+import org.mongodb.kbson.BsonObjectId
+import org.mongodb.kbson.BsonString
+import org.mongodb.kbson.BsonValue
+import kotlin.reflect.KClass
+import kotlin.test.Test
+import kotlin.test.assertContentEquals
+import kotlin.test.assertEquals
+import kotlin.test.assertNull
+
+class BsonEncoderTests {
+    @kotlinx.serialization.Serializable
+    class SerializableClass
+
+    private val primitiveValues: List<Pair<Any?, BsonValue>> = listOf(
+        10.toByte() to BsonInt32(10.toByte().toInt()),
+        20.toShort() to BsonInt32(20.toShort().toInt()),
+        30 to BsonInt32(30),
+        40L to BsonInt64(40L),
+        50F to BsonDouble(50F.toDouble()),
+        2.0 to BsonDouble(2.0),
+        true to BsonBoolean.TRUE,
+        "hello world" to BsonString("hello world"),
+        'c' to BsonString('c'.toString()),
+        byteArrayOf(0x00, 0x01, 0x03) to BsonBinary(
+            BsonBinarySubType.BINARY,
+            byteArrayOf(0x00, 0x01, 0x03)
+        ),
+        null to BsonNull
+    )
+
+    private val realmValues = listOf(
+        MutableRealmInt.create(15) to BsonInt64(15),
+        RealmUUID.from("ffffffff-ffff-ffff-ffff-ffffffffffff").let {
+            it to BsonBinary(BsonBinarySubType.UUID_STANDARD, it.bytes)
+        },
+        ObjectId.create().let {
+            it to BsonObjectId((it as ObjectIdImpl).bytes)
+        },
+        RealmInstant.from(
+            epochSeconds = 1668425451,
+            nanosecondAdjustment = 862000000
+        ) to BsonDateTime(1668425451862)
+    )
+
+    private val listValue: Pair<List<Any?>, BsonArray> =
+        primitiveValues.map { it.first } to BsonArray(primitiveValues.map { it.second })
+
+    private val mapValue =
+        primitiveValues.mapIndexed { index, pair ->
+            index.toString() to pair.first
+        }.toMap() to BsonDocument(
+            primitiveValues.mapIndexed { index, pair ->
+                index.toString() to pair.second
+            }.toMap()
+        )
+
+    @Test
+    fun encodeToString() {
+        (primitiveValues + realmValues + listValue + mapValue).forEach { (value, bsonValue) ->
+            assertEquals(
+                bsonValue,
+                BsonEncoder.encodeToBsonValue(value)
+            )
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    internal fun serializer(value: KClass<*>): KSerializer<*> = when (value) {
+        ObjectIdImpl::class -> BsonEncoder.serializersModule.serializer<ObjectId>()
```

Comment:
It looks easy to let this function get out of sync with changes to the codebase, but not sure what we can do about it ? 🤔

Reply:
Yeah, I am not sure how we could get it tracked.